Seagate/cortx-manager
# EOS-22627: Openldap configuration in CSM-setup

# Backend

# Problem Statement
[EOS-22627: F-45J: Cortx User Schema Configuration in Openldap (Config Phase)](https://jts.seagate.com/browse/EOS-22627)
- _Overview: JIRA number/GitHub Issue added to PR: [Y/N]:_ Y
- _Describe the problem this patch intends to solve._ 
 This patch will configure the openldap for cortxusers in CSM-setup

## Design
Ldap configuration in CSM-Setup,

config phase- 

1. push cortxuser schema(attributes and objectClasses) in openldap - cortxuser.ldif

2. initialization of dc=csm,dc=seagate,dc=com - csm-ldap-init.ldif

3. give permissions to sgiamadmin user for reading, writing

4. create organization account , o=userAccount - userAccount.ldif

Reset phase- 

1. delete all the cortxuser accounts created under organization account (o=userAccount)

cleanup phase - 

1. delete cortxuser.ldif from /etc/openldap/slapd.d/cn=config/cn=schema/
2. revoke the permissions given in config phase
3. restart slapd service


## Coding

- _Coding conventions are followed and code is consistent. Yes/No?_ 
- _Confirm All CODACY errors are resolved. Yes/No?_

## Testing

- _Confirm that Test Cases are added (for both the cases, fix and feature). Yes/No?_
- _Confirm Test Cases cover Happy Path, Non-Happy Path and Scalability. Yes/No?_
- _Confirm Testing was performed with installed RPM. Yes/No?_
- _Confirm all the ut/regression/smoke test has been performed [Y/N]_

1. csm_setup post_install, prepare, config, init 
![image](https://user-images.githubusercontent.com/36843912/128660919-3db34332-c960-4452-8551-470ea4eeecaa.png)

2. Check if, cortxuser schema is pushed and appropriate permissions has been set.
![image](https://user-images.githubusercontent.com/36843912/128661052-4516e057-af06-4907-b977-e8da878b2f37.png)

3. Add sample cortxusers in ldap
![image](https://user-images.githubusercontent.com/36843912/128661342-f3528d89-8f75-456b-b490-e35c5e68d2d2.png)

4. Fetch sample users
![image](https://user-images.githubusercontent.com/36843912/128661483-d7803c45-035e-4622-8a7d-454a7b38e652.png)

![image](https://user-images.githubusercontent.com/36843912/128661535-a9848de5-0ee8-4723-86e9-260d170b5967.png)

5. Reset phase, check if sample users are deleted.
![image](https://user-images.githubusercontent.com/36843912/128661642-8037ce07-c407-41d0-9fb8-2d2eead6408a.png)

6. Cleanup phase, check if cortxuser schema is deleted and permissions are revoked.
![image](https://user-images.githubusercontent.com/36843912/128661786-9e25e85e-8a47-4986-bbcd-8e44fbf364a9.png)

7. Stop cluster, perform reset, start cluster. After cluster is up all is working fine.

> [root@ssc-vm-rhev4-0918 ~]# vi /opt/seagate/cortx/csm/conf/reset.py
[root@ssc-vm-rhev4-0918 ~]# clear
[root@ssc-vm-rhev4-0918 ~]# pcs cluster stop --force
Stopping Cluster (pacemaker)...
Stopping Cluster (corosync)...
[root@ssc-vm-rhev4-0918 ~]# csm_setup reset --config json:///opt/seagate/cortx_configs/provisioner_cluster.json
:PASS
[root@ssc-vm-rhev4-0918 ~]# pcs cluster start
Starting Cluster (corosync)...
Starting Cluster (pacemaker)...
[root@ssc-vm-rhev4-0918 ~]# 

8. stop the cluster, perform reset, cleanup, post_install, prepare, config, init phases. start the cluster. All is working fine.

> [root@ssc-vm-rhev4-0918 ~]# pcs cluster stop --force
Stopping Cluster (pacemaker)...
Stopping Cluster (corosync)...
[root@ssc-vm-rhev4-0918 ~]# csm_setup reset --config json:///opt/seagate/cortx_configs/provisioner_cluster.json
:PASS
[root@ssc-vm-rhev4-0918 ~]# csm_setup cleanup --config json:///opt/seagate/cortx_configs/provisioner_cluster.json
:PASS
[root@ssc-vm-rhev4-0918 ~]# csm_setup post_install --config json:///opt/seagate/cortx_configs/provisioner_cluster.json
:PASS
[root@ssc-vm-rhev4-0918 ~]# csm_setup prepare --config json:///opt/seagate/cortx_configs/provisioner_cluster.json
:PASS
[root@ssc-vm-rhev4-0918 ~]# csm_setup config --config json:///opt/seagate/cortx_configs/provisioner_cluster.json
:PASS
[root@ssc-vm-rhev4-0918 ~]# csm_setup init --config json:///opt/seagate/cortx_configs/provisioner_cluster.json
:PASS
[root@ssc-vm-rhev4-0918 ~]# pcs cluster start
Starting Cluster (corosync)...
Starting Cluster (pacemaker)...
[root@ssc-vm-rhev4-0918 ~]#


## Checklist

- _PR is self-reviewed? Yes/No?_ Yes
- _GitHub Issue is updated_ Yes
- _Jira is updated_ Yes
  -  _Check if the description is clear and explained._ 
    -  _Check Acceptance Criterion is defined._
      -  _All the tests performed should be mentioned before Resolving a JIRA._
        -  _Verification needs to be done before marked as Closed/Verified_
        - _Any interface change Yes/No?_
        - _Change notified to other gatekeepers: Yes/No?_
        - _Code reviews done and addressed: Yes/No?_
        - _Codacy issues reviewed and addressed: Yes/No?_
        - _Deployment test : Done/Not?_ Done
        - _UT/smoke test: Done/Not?_ Done
        - _Jira number added to PR: Yes/No?_ Yes
        - _Github merge done using UI: Yes/No?_
        - _Side effects on other features - deployment/update/node-replacement_
        - _Changes to quick-start-guide/community impact_
        - _All dependent component code PR are raised or already merged Yes/No_
        - _All dependent code already merged in landing branch Yes/No_
